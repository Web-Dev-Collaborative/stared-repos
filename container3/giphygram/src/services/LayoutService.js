// Set of screen sizes that app is supporting.
export const SCREEN_SIZES = {
  xl: { id: 'xl', mediaQuery: '(min-width: 1200px)' },
  lg: { id: 'lg', mediaQuery: '(max-width: 1200px) and (min-width: 960px)' },
  md: { id: 'md', mediaQuery: '(max-width: 950px) and (min-width: 720px)' },
  sm: { id: 'sm', mediaQuery: '(max-width: 720px) and (min-width: 540px)' },
  xs: { id: 'xs', mediaQuery: '(max-width: 540px)' },
};

// Supported number of columns in 12-column templates.
export const SUPPORTED_COLUMNS_NUMS = [1, 2, 3, 4, 6, 12];

export const DEFAULT_COLUMNS_NUM = 3;

export const TOTAL_NUMBER_OF_LAYOUT_COLUMNS = 12;

export class LayoutService {
  /**
   * Get default number of columns for 12-column template depending on screen size.
   * @param {string} screenSize
   * @return {number}
   */
  static screenSizeToColumns(screenSize) {
    switch (screenSize) {
      case SCREEN_SIZES.xs.id:
        return 1;

      case SCREEN_SIZES.sm.id:
      case SCREEN_SIZES.md.id:
        return 3;

      case SCREEN_SIZES.lg.id:
      case SCREEN_SIZES.xl.id:
        return 4;

      default:
        return DEFAULT_COLUMNS_NUM;
    }
  }

  /**
   * Chooses the number of layout columns from manual and automatic one.
   * @param {?number} responsiveColumnsNum - number of layout columns according to window size
   * @param {?number} manualColumnsNum - number of layout columns according to user selection.
   * @return {number}
   */
  static choseColumnsNum(responsiveColumnsNum, manualColumnsNum) {
    // Manual column selection has higher priority than automatic columns number
    // calculated based on window size.
    return manualColumnsNum || responsiveColumnsNum;
  }
}
