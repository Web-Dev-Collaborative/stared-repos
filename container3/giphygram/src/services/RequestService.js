import axios from 'axios';
import { REQUEST_TIMEOUT } from '../config';

// RequestService is created as a wrapper on top of axios. This is done in order to be able to
// switch to new http-requests library if needed. Utilizing interfaces might be a good option here.
export class RequestService {
  // Perform GET request with parameters.
  static get(url, params = {}) {
    return axios.get(
      url,
      { ...this.getDefaultAxiosConfig(), params },
    );
  }

  // Generate default Axios configuration for requests.
  // We may put here default timeouts etc.
  static getDefaultAxiosConfig() {
    return {
      timeout: REQUEST_TIMEOUT,
    };
  }
}
