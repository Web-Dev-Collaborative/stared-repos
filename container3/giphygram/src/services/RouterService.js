const SEARCH_QUERY_PARAM = 'query';

export class RouterService {
  constructor(history, location) {
    this.history = history;
    this.location = location;
  }

  pushSearchQuery(query = '') {
    this.history.push({
      search: query ? `?${SEARCH_QUERY_PARAM}=${query}` : '',
    });
  }

  getSearchQuery() {
    const searchParams = new URLSearchParams(this.location.search);
    return searchParams.get(SEARCH_QUERY_PARAM);
  }
}
