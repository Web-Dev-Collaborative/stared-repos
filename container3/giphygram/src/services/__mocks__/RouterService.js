export class RouterService {
  pushSearchQuery = jest.fn();

  getSearchQuery = jest.fn();
}
