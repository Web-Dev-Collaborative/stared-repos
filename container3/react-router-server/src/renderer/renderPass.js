import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
