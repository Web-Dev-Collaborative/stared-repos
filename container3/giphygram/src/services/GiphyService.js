import { RequestService } from './RequestService';
import { GIPHY_API_HOST, GIPHY_API_KEY } from '../config';
import { defaultSearchParams } from '../reducers/searchParamsReducer';

// GiphyService is responsible for all integrations with Giphy API.
export class GiphyService {
  // Search for GIFs by specific search query.
  static search({
    query = defaultSearchParams.query,
    limit = defaultSearchParams.limit,
    offset = defaultSearchParams.offset,
    rating = defaultSearchParams.rating,
    lang = defaultSearchParams.lang,
  } = {}) {
    const searchParams = {
      api_key: GIPHY_API_KEY,
      q: query,
      limit,
      offset,
      rating,
      lang,
    };

    return RequestService.get(
      `${GIPHY_API_HOST}/v1/gifs/search`,
      searchParams,
    );
  }
}
