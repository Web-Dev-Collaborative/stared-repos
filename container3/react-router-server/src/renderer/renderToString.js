import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
