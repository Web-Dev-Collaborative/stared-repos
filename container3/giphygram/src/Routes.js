import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Route, Switch, withRouter } from 'react-router-dom';
import { SearchPage } from './components/searchPage/SearchPage';
import { RouterService } from './services/RouterService';
import { updateSearchQuery } from './actions/searchParamsActions';
import { search } from './actions/searchActions';

export class Routes extends React.Component {
  static propTypes = {
    routerService: PropTypes.instanceOf(RouterService).isRequired,
    updateSearchQuery: PropTypes.func.isRequired,
    search: PropTypes.func.isRequired,
  };

  componentDidMount() {
    const {
      routerService,
      updateSearchQuery: searchQueryFromLocationCallback,
      search: searchCallback,
    } = this.props;

    // Check if search query has been submitted through the URL.
    // In case if search query is in URL we need to launch the search.
    const searchQueryFromLocation = routerService.getSearchQuery();
    if (searchQueryFromLocation) {
      // Update search form parameters.
      searchQueryFromLocationCallback(searchQueryFromLocation);
      // Launch the search and populate the state with search results.
      searchCallback({ query: searchQueryFromLocation });
    }
  }

  render() {
    // Currently we have only one route. But the next step of the App development might be to create
    // a dedicated pages for each GIF with additional details. Or to display most trending GIFs
    // on the /home and search results on the /search page.
    return (
      <Switch>
        <Route path="/" component={SearchPage} />
      </Switch>
    );
  }
}

const mapStateToProps = (state, props) => ({
  routerService: new RouterService(props.history, props.location),
});

const mapDispatchToProps = {
  updateSearchQuery,
  search,
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Routes),
);
