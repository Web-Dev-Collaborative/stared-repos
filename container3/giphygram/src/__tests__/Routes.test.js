import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Routes } from '../Routes';
import { RouterService } from '../services/RouterService';

jest.mock('../services/RouterService');
jest.mock('../components/searchPage/SearchPage', () => ({
  SearchPage: () => 'SearchPage',
}));

describe('Routes', () => {
  it('should not launch the search when search param is empty in URL', () => {
    const searchMock = jest.fn();
    const updateSearchQueryMock = jest.fn();
    const routerService = new RouterService(null, null);

    const component = renderer.create((
      <MemoryRouter>
        <Routes
          routerService={routerService}
          updateSearchQuery={updateSearchQueryMock}
          search={searchMock}
        />
      </MemoryRouter>
    ));

    expect(searchMock).not.toHaveBeenCalled();
    expect(updateSearchQueryMock).not.toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should launch the search when search param is not empty in URL', () => {
    const searchQuery = 'kitten';
    const searchMock = jest.fn();
    const updateSearchQueryMock = jest.fn();
    const routerService = new RouterService(null, null);
    routerService.getSearchQuery.mockImplementation(() => searchQuery);

    const component = renderer.create((
      <MemoryRouter>
        <Routes
          routerService={routerService}
          updateSearchQuery={updateSearchQueryMock}
          search={searchMock}
        />
      </MemoryRouter>
    ));

    expect(searchMock).toHaveBeenCalledTimes(1);
    expect(searchMock).toHaveBeenCalledWith({ query: searchQuery });

    expect(updateSearchQueryMock).toHaveBeenCalledTimes(1);
    expect(updateSearchQueryMock).toHaveBeenLastCalledWith(searchQuery);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
