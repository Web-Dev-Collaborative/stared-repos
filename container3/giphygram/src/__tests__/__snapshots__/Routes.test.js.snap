// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes should launch the search when search param is not empty in URL 1`] = `"SearchPage"`;

exports[`Routes should not launch the search when search param is empty in URL 1`] = `"SearchPage"`;
